relays/finality-relay/src/finality_loop.rs

```diff
@@ -0,0 +1,558 @@
+		if let Some(justified_header) = justified_header {
+			recent_finality_proofs.clear();
+			selected_finality_proof = Some((justified_header, finality_proof));
+		} else {
+			// the number of proofs read during single wakeup is expected to be low, so we aren't pruning
+			// `recent_finality_proofs` collection too often
+			recent_finality_proofs.push((finality_proof_target_header_number, finality_proof));
+		}
+	}
+
+	// remove obsolete 'recent' finality proofs + keep its size under certain limit
+	let oldest_finality_proof_to_keep = selected_finality_proof
+		.as_ref()
+		.map(|(header, _)| header.number())
+		.unwrap_or(best_number_at_target);
+	prune_recent_finality_proofs::<P>(
+		oldest_finality_proof_to_keep,
+		recent_finality_proofs,
+		sync_params.recent_finality_proofs_limit,
+	);
+
+	Ok(selected_finality_proof)
+}
+
+pub(crate) fn prune_unjustified_headers<P: FinalitySyncPipeline>(
+	justified_header_number: P::Number,
+	unjustified_headers: &mut UnjustifiedHeaders<P>,
+) -> Option<P::Header> {
+	prune_ordered_vec(justified_header_number, unjustified_headers, usize::MAX, |header| {
+		header.number()
+	})
+}
+
+pub(crate) fn prune_recent_finality_proofs<P: FinalitySyncPipeline>(
+	justified_header_number: P::Number,
+	recent_finality_proofs: &mut FinalityProofs<P>,
+	recent_finality_proofs_limit: usize,
+) {
+	prune_ordered_vec(
+		justified_header_number,
+		recent_finality_proofs,
+		recent_finality_proofs_limit,
+		|(header_number, _)| *header_number,
+	);
+}
+
+fn prune_ordered_vec<T, Number: relay_utils::BlockNumberBase>(
+	header_number: Number,
+	ordered_vec: &mut Vec<T>,
+	maximal_vec_size: usize,
+	extract_header_number: impl Fn(&T) -> Number,
+) -> Option<T> {
+	let position = ordered_vec.binary_search_by_key(&header_number, extract_header_number);
+
+	// first extract element we're interested in
+	let extracted_element = match position {
+		Ok(position) => {
+			let updated_vec = ordered_vec.split_off(position + 1);
+			let extracted_element = ordered_vec.pop().expect(
+				"binary_search_by_key has returned Ok(); so there's element at `position`;\
+					we're splitting vec at `position+1`; so we have pruned at least 1 element;\
+					qed",
+			);
+			*ordered_vec = updated_vec;
+			Some(extracted_element)
+		}
+		Err(position) => {
+			*ordered_vec = ordered_vec.split_off(position);
+			None
+		}
+	};
+
+	// now - limit vec by size
+	let split_index = ordered_vec.len().saturating_sub(maximal_vec_size);
+	*ordered_vec = ordered_vec.split_off(split_index);
+
+	extracted_element
+}
+
+fn print_sync_progress<P: FinalitySyncPipeline>(
+	progress_context: (Instant, Option<P::Number>),
+	best_number_at_source: P::Number,
+	best_number_at_target: P::Number,
+) -> (Instant, Option<P::Number>) {
+	let (prev_time, prev_best_number_at_target) = progress_context;
+	let now = Instant::now();
+
+	let need_update = now - prev_time > Duration::from_secs(10)
```

Comment:
```suggestion
	let need_update = prev_time.elapsed() > Duration::from_secs(10)
```
?

Reply:
We'll need `now` anyways - it is a part of `progress_context`. So I'm just reusing the same `now` value.